Initial screen + Sign up/in

TahaKamil opened the pull request:
Resolves: #136, #134, #148

**Description**

What did you change? 
* updated the UI of the First screen and the sign in/up screens to reflect the figma
* required sign-up to have unique usernames
* **_Tests for UI will be done later, I wanted everybody to have the firebase changes before the meeting_**

How should a reviewer test? 
You will the new firebase ruleset:
```
rules_version = '2';

service cloud.firestore {
  match /databases/{database}/documents {
    // can create user if and only if they have a valid user (which involves having a valid username)
    match /users/{userId} {
        allow create: if request.auth.uid == userId && isUniqueUsernameAndOwner(request.resource.data.username);
    }

    // makes sure that only unique usernames can be added to usernames
    match /usernames/{username} {
      allow create: if isUniqueUsernameAndOwner(username);
      allow read;
    }

    // Function to ensure a user can create an account with a unique username
    function isUniqueUsernameAndOwner(username) {
      let isOwner = request.auth.uid == request.resource.data.uid;

      // Check if username already exists in the database
      let usernameNotExists = !exists(/databases/$(database)/documents/usernames/$(username));

      return isOwner && usernameNotExists;
    }
  }
}
```

**Reviewer Checklist**

- [x] Are all relevant issues referenced by this PR?
- [ ] Does the app still build?
- [x] Do all tests pass? (TODO: Automate this with GitHub actions)
- [ ] Does the feature work as expected?
- [ ] Does the feature still keep the app bug free?
- [ ] Is the description of the changes correct/present?

Commit:
Finished Inital Screen UI

The buttons already worked, so I simply had to clean up the UI and make
it so that it is responsive to screen size.

Co-authored-by: Aidan <PossiblyAShrub@users.noreply.github.com>

Commit:
Finished UI and Removed Key

I accidentally leaked google-services.json, so I added it into the
.gitignore and also regenerated a key. I am also finished the UI of the
Intial Screen, Sign-In and Sign-Up.

Commit:
Merge branch 'main' of github.com:cmput301-w25/project-kernelcrew into intial-screen

Commit:
Done UI and Associated Login Classes

I have finished all the inital screen stuff and only need to do the
test, but right now the tests are being a pain, so I will leave them for
now.

Commit:
Merge branch 'main' of github.com:cmput301-w25/project-kernelcrew into intial-screen

Commit:
final fixes

Commit:
Fixed Test

Fixed the test so that I could pass it and now I will do a PR. Later on
implemnting the tests for each and every UI element etc.

Commit:
Last minute fixes

PossiblyAShrub commented on firestore.rules:
```diff
@@ -2,8 +2,25 @@ rules_version = '2';
   match /databases/{database}/documents {
-    match /{document=**} {
-      allow read, write: if request.auth != null;
+    // can create user if and only if they have a valid user (which involves having a valid username)
```
This prevents us from creating new moods. That's why the test in CI is failing.

TahaKamil replied:
Good catch! Let me fix that real quick

Commit:
Think I Fixed Test

The test passed like once, so let me give it a shot.

Commit:
Home navigation extension

Commit:
Added s to moodEvents

TahaKamil commented:
Please delete the branch after approving and merging.

PossiblyAShrub commented on code/app/src/androidTest/java/com/kernelcrew/moodapp/CreateMoodEventTest.java:
```diff
@@ -31,16 +36,16 @@
 
-    @BeforeClass
-    public static void login() throws ExecutionException, InterruptedException {
+    @Before
```
If we left this as `@BeforeClass`, then we wouldn't need the "is user already logged in" check.

PossiblyAShrub commented on code/app/src/androidTest/java/com/kernelcrew/moodapp/CreateMoodEventTest.java:
```diff
@@ -31,16 +36,16 @@
+        if (auth.getCurrentUser() == null) {
+            Tasks.await(auth.createUserWithEmailAndPassword("test@toasted.com", "password"));
+        }
+        scenario = ActivityScenario.launch(MainActivity.class);
```
Why do we need this instead of using `@Rule` like above?

TahaKamil replied:
> Why do we need this instead of using `@Rule` like above?

The problem is that the rule gets called before the we create a Authenticated user. This leads to the app being launched at the Initial screen instead of the home screen.

PossiblyAShrub replied:
If we make this static and use `@BeforeClass`, that won't be an issue. Additionally we won't have to manually relaunch the activity scenario every time.

TahaKamil replied:
Okay, let me change some things around then and try and get it running.

PossiblyAShrub commented on code/app/src/androidTest/java/com/kernelcrew/moodapp/HomeFeedNavigationTest.java:
```diff
@@ -61,11 +69,19 @@ public void testNavigationToHomeFeed() {
+    private void createUser(String email, String password) throws InterruptedException {
+        FirebaseAuth auth = FirebaseAuth.getInstance();
+        Task<AuthResult> createUserTask = auth.createUserWithEmailAndPassword(email, password);
+        while (!createUserTask.isComplete()) {
```
We should use `Tasks.await` instead.

PossiblyAShrub commented on code/app/src/main/res/layout/fragment_auth_sign_in.xml:
```diff
@@ -1,86 +1,144 @@
+                app:startIconTint="@null"
+                app:startIconContentDescription="@string/prompt_password"
+                app:endIconMode="password_toggle"
+                app:endIconDrawable="@drawable/ic_visibility"
```
Can we toggle this to an eye with a cross through it when the password is already visible. Right now the UI feels less smooth without it...

<img width="253" alt="image" src="https://github.com/user-attachments/assets/d06a574a-d850-4df6-b5fa-cdd5a8fb3291" />
<img width="253" alt="Screenshot 2025-03-05 at 10 35 46" src="https://github.com/user-attachments/assets/cfcdbe8e-48c1-46b4-85f9-74045f464bd6" />

PossiblyAShrub commented on code/app/build.gradle:
```diff
@@ -41,7 +41,8 @@ dependencies {
     implementation(platform(libs.firebase.bom))
-    implementation 'com.google.firebase:firebase-firestore'
+    implementation libs.firebase.firestore
+    implementation libs.play.services.auth
```
Do we use this dependency anywhere? What happens if we remove it? It seems to work fine without in my testing.

TahaKamil replied:
I was trying to resolve the warning/error i was getting. It was something something .gms and I forgot to remove that line.

PossiblyAShrub replied:
Ah okay. I think @adityadipakpatel ran into this too. Feel free to leave it if it brings the warning/error back.

TahaKamil replied:
It did not really fix it. I think google play store runs a security check of the app automatically and when you fail it they seem to revoke the package, thats what I believe is happening?

this might help @adityadipakpatel if it is causing him problems 
https://stackoverflow.com/questions/15309694/where-is-com-google-android-gms

PossiblyAShrub reviewed (dismissed):
I noticed we aren't showing errors on wrong password/email or duplicate email/username.

We also don't show errors for:
- Bad password (firebase does some validation like >6 characters)
- Bad email (not a valid email format)

Nor does it seem that the validation for duplicate usernames is working.

We should also add tests for all of these cases.

Commit:
@before -> @beforeClass

Commit:
Requested Changes

Made the eye in the password toggle in visibility states, changed the
gradle file so that it is back up to date again without the gms solution
attempt.

Commit:
Merge branch 'main' of github.com:cmput301-w25/project-kernelcrew into intial-screen

PossiblyAShrub commented on code/app/src/androidTest/java/com/kernelcrew/moodapp/HomeFeedNavigationTest.java:
```diff
@@ -7,48 +7,47 @@
-        SystemClock.sleep(2000);
+    public void setUp() throws InterruptedException, ExecutionException {
+        createUser(TEST_EMAIL, TEST_PASSWORD);
+        SystemClock.sleep(1000);
```
Instead of sleeping, you can call `Tasks.await` on the `createUserTask`. That should be much less flaky -- and, in this case, faster.

PossiblyAShrub reviewed (commented):
The sign-in is also missing the two eye icon states.

Commit:
Finally Done

I remade tests and it is fianlly done ;w;. No wonder ppl hate tests with
such a passion, the tests actually fail more then my UI itself
does...like how does that make any sense??

Commit:
Merge branch 'main' of github.com:cmput301-w25/project-kernelcrew into intial-screen

Commit:
Fixes to Tests

Commit:
Added Elavation to Sign Up Button

Commit:
Elavation for Sign In

I forgot and I wanted to keep it even for both buttons.

TahaKamil commented:
10 minute CI for commit "Fixes to Tests" is diabolical...I am finally understanding the pain of tests 😭

TahaKamil commented:
@PossiblyAShrub do you know why the tests are failing? I think it has something to do with the small device but I am unsure how to fix it.

Commit:
Update ID's for Images
Signed-off-by: “Taha <“tkamil@ualberta.ca”>

Commit:
Merge branch 'main' of github.com:cmput301-w25/project-kernelcrew into intial-screen

Commit:
merge errors

Commit:
All the Tests Passed

LETS GOOOOOOOOOOO!

Commit:
Fixes

Commit:
Test Fixes Again

TahaKamil commented:
DO NOT MERGE, still fixing stuff

Commit:
Final Pushes Before Merge

I fixed some bugs and tested the features of the intial screen etc.

Commit:
More QOL

I fixed some minor bigs with errors and generally made it a better
experince.

Commit:
Fixes

Remidied app crashes and fixed it.
